src/orange/detail/BIHBuilder.cc

```diff
@@ -0,0 +1,228 @@
+
+    centers_.resize(bboxes_.size());
+    std::transform(bboxes_.begin(),
+                   bboxes_.end(),
+                   centers_.begin(),
+                   [&](BoundingBox const& bbox) { return center(bbox); });
+
+    partitioner_ = BIHPartitioner(&bboxes_, &centers_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create BIH Nodes.
+ */
+BIHParams BIHBuilder::operator()() const
+{
+    VecIndices indices;
+    VecIndices inf_volids;
+
+    for (auto i : range(bboxes_.size()))
+    {
+        LocalVolumeId id(i);
+
+        if (!is_infinite(bboxes_[i]))
+        {
+            indices.push_back(id);
+        }
+        else
+        {
+            inf_volids.push_back(id);
+        }
+    }
+
+    VecNodes nodes;
+    this->construct_tree(indices, nodes, static_cast<BIHNodeId>(-1));
+
+    auto [inner_nodes, leaf_nodes] = this->arrange_nodes(std::move(nodes));
+
+    BIHParams params;
+
+    params.bboxes = ItemMap<LocalVolumeId, BoundingBoxId>(
+        make_builder(bbox_storage_).insert_back(bboxes_.begin(), bboxes_.end()));
+
+    params.inner_nodes
+        = make_builder(inner_node_storage_)
+              .insert_back(inner_nodes.begin(), inner_nodes.end());
+    params.leaf_nodes = make_builder(leaf_node_storage_)
+                            .insert_back(leaf_nodes.begin(), leaf_nodes.end());
+
+    params.inf_volids = make_builder(lvi_storage_)
+                            .insert_back(inf_volids.begin(), inf_volids.end());
+
+    return params;
+}
+
+//---------------------------------------------------------------------------//
+// HELPER FUNCTIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Recursively construct BIH nodes for a vector of bbox indices.
+ */
+void BIHBuilder::construct_tree(VecIndices const& indices,
+                                VecNodes& nodes,
+                                BIHNodeId parent) const
+{
+    auto current_index = nodes.size();
+    nodes.resize(nodes.size() + 1);
+
+    if (partitioner_.is_partitionable(indices))
+    {
+        BIHInnerNode node;
+        node.parent = parent;
+
+        auto p = partitioner_(indices);
+        auto ax = to_int(p.axis);
+
+        BIHInnerNode::BoundingPlane left_plane{
+            p.axis,
+            static_cast<BIHInnerNode::location_type>(p.left_bbox.upper()[ax])};
+
+        BIHInnerNode::BoundingPlane right_plane{
+            p.axis,
+            static_cast<BIHInnerNode::location_type>(p.right_bbox.lower()[ax])};
+
+        node.bounding_planes = {left_plane, right_plane};
+
+        // Recursively construct the left and right branches
+        node.children[BIHInnerNode::Edge::left] = BIHNodeId(nodes.size());
+        this->construct_tree(p.left_indices, nodes, BIHNodeId(current_index));
+
+        node.children[BIHInnerNode::Edge::right] = BIHNodeId(nodes.size());
+        this->construct_tree(p.right_indices, nodes, BIHNodeId(current_index));
```

Comment:
if you make a `Partition::indices` array this becomes a loop over `range(Edge::size_)`